### PR TITLE
Add get latest service function

### PR DESCRIPTION
**Pull request by cortadocodes**

### CODE REVIEW

File: octue/cloud/service_id.py [hunk 1]
```diff
@@ -162,3 +163,45 @@ def convert_service_id_to_pub_sub_form(service_id):
...
 
     return service_id
+
+
+def split_service_id(service_id):
+    """Split a service ID into its namespace, name, and, if present, its revision tag.
+
+    :param str service_id: a service ID
+    :return tuple(str, str, str|None): the namespace, name, and revision tag
+    """
```
Comment (thclark) [inline octue/cloud/service_id.py:173]: `split_service_id` appears to be able to return a blank string as a revision tag, say if I gave it `octue/service:`. I think that edge case should be handled to always return `None`

File: tests/cloud/pub_sub/test_service.py [hunk 2]
```diff
@@ -71,14 +71,14 @@ def test_serve_fails_if_service_with_same_id_already_exists(self):
                 side_effect=google.api_core.exceptions.AlreadyExists(""),
             ):
                 with self.assertRaises(exceptions.ServiceAlreadyExists):
-                    MockService(backend=BACKEND, service_id="my-org/existing-service:latest").serve()
+                    MockService(backend=BACKEND, service_id="my-org/existing-service:2.3.0").serve()
```
Comment (thclark) [inline tests/cloud/pub_sub/test_service.py:74]: I'd make `2.3.0` a constant, say `MOCK_SERVICE_REVISION_TAG`

File: tests/cloud/test_service_id.py [hunk 3]
```diff
@@ -162,3 +172,70 @@ def test_no_error_raised_if_service_sruid_components_valid(self):
         ):
             with self.subTest(namespace=namespace, name=name, revision_tag=revision_tag):
                 validate_service_sruid(namespace=namespace, name=name, revision_tag=revision_tag)
+
+
+class TestSplitServiceID(unittest.TestCase):
+    def test_split_sruid(self):
+        """Test that a valid SRUID can be split into its namespace, name, and revision tag."""
+        namespace, name, revision_tag = split_service_id("octue/my-service:2.3.0")
```
Comment (thclark) [inline tests/cloud/test_service_id.py:180]: Add tests for invalid values like `octue/my-service:`, `myservice:1.2.0`, `/myservice`, `/myservice:` and `myservice`. I commented above that `octue/my-service:` should return `None` as a revision tag, but actually I think all of these should raise an exception inheriting from ValueError

File: tests/cloud/test_service_id.py [hunk 4]
```diff
@@ -162,3 +172,70 @@ def test_no_error_raised_if_service_sruid_components_valid(self):
...
+        mock_response._content = json.dumps({"revision_tag": "1.3.9"}).encode()
+
+        with patch("requests.get", return_value=mock_response):
+            latest_sruid = get_latest_service_revision_sruid(
+                namespace="my-org",
+                name="my-service",
+                service_registries=self.SERVICE_REGISTRIES,
+            )
+
+        self.assertEqual(latest_sruid, "my-org/my-service:1.3.9")
```
Comment (thclark) [inline tests/cloud/test_service_id.py:222]: Yep, this is perfect

File: octue/cloud/pub_sub/service.py [hunk 5]
```diff
@@ -50,10 +57,11 @@ class Service:
     :param str|None service_id: a unique ID to give to the service (any string); a UUID is generated if none is given
     :param callable|None run_function: the function the service should run when it is called
     :param str|None name: an optional name to use for the service to override its ID in its string representation
+    :param iter(dict)|None service_registries: the names and endpoints of the registries used to resolve service revisions when asking questions; these should be in priority order (highest priority first)
     :return None:
     """
 
-    def __init__(self, backend, service_id=None, run_function=None, name=None):
+    def __init__(self, backend, service_id=None, run_function=None, name=None, service_registries=None):
```
Comment (thclark) [inline octue/cloud/pub_sub/service.py:64]: I believe that as we move forward, the `backend` here should be able to be initialised from the service registry response.

**Review by thclark — CHANGES_REQUESTED** [review]
A couple of extremely minor comments on the code, which is exactly what we're looking for right now.
Biggest issue is that documentation is totally missing. How do I now set up a service to use a service registry?

**Comment by cortadocodes** [comment]
@thclark literally just pushed the docs 😂

File: octue/cloud/service_id.py (hunk 1 above)
Comment (cortadocodes) [reply octue/cloud/service_id.py:173]: Good point

File: octue/cloud/pub_sub/service.py (hunk 5 above)
Comment (cortadocodes) [reply octue/cloud/pub_sub/service.py:64]: Yep I agree. Shall I leave any implementation of that until registries support that?

Comment (thclark) [reply octue/cloud/pub_sub/service.py:64]: Yes, absolutely leave that for now.

File: octue/cloud/service_id.py [hunk 6]
```diff
@@ -162,3 +163,45 @@ def convert_service_id_to_pub_sub_form(service_id):
...
+    try:
+        name, revision_tag = name_and_revision_tag.split(":")
+    except ValueError:
+        name = name_and_revision_tag
+        revision_tag = None
+
+    return namespace, name, revision_tag
+
+
+def get_latest_service_revision_sruid(namespace, name, service_registries):
```
Comment (thclark) [inline octue/cloud/service_id.py:185]: this could just be called `get_latest_sruid` because it currently says "get latest service revision service revision unique identifier"

**Review by thclark — COMMENTED** [review]
Extra notes on docs

File: octue/cloud/service_id.py (hunk 6 above)
Comment (cortadocodes) [reply octue/cloud/service_id.py:185]: I was just thinking the same myself
